src/agreement/mod.rs

```diff
@@ -387,53 +459,36 @@ impl<NodeUid: Clone + Debug + Eq + Hash + Ord> Agreement<NodeUid> {
         self.received_aux.clear();
         self.received_conf.clear();
         self.conf_round = false;
+        self.conf_vals = BinValues::None;
         self.epoch += 1;
+        let nonce = Nonce::new(
+            self.netinfo.invocation_id().as_ref(),
+            self.hb_epoch,
+            self.epoch,
+        );
+        self.common_coin = CommonCoin::new(self.netinfo.clone(), nonce);
         debug!(
             "Agreement instance {:?} started epoch {}",
             self.netinfo.our_uid(),
             self.epoch
         );
     }
+}
 
-    /// Gets a common coin and uses it to compute the next decision estimate and outputs the
-    /// optional decision value.  The function may start the next epoch. In that case, it also
-    /// returns a message for broadcast.
-    fn invoke_coin(&mut self, vals: BinValues) -> AgreementResult<()> {
-        debug!(
-            "{:?} invoke_coin in epoch {}",
-            self.netinfo.our_uid(),
-            self.epoch
-        );
-        // FIXME: Implement the Common Coin algorithm. At the moment the
-        // coin value is common across different nodes but not random.
-        let coin = (self.epoch % 2) == 0;
-
-        // Check the termination condition: "continue looping until both a
-        // value b is output in some round r, and the value Coin_r' = b for
-        // some round r' > r."
-        self.terminated = self.terminated || self.decision == Some(coin);
-        if self.terminated {
-            return Ok(());
-        }
+#[derive(Clone)]
+struct Nonce(Vec<u8>);
 
-        self.start_next_epoch();
+impl Nonce {
+    pub fn new(invocation_id: &[u8], hb_epoch: u64, agreement_epoch: u32) -> Self {
+        Nonce(Vec::from(format!(
+            "Nonce for Honey Badger {:?} @ epoch {}:{}",
```

Comment:
It would be slightly more efficient to just use `"{:?}@{}:{}"` here (and then the proposer ID, too, of course).